feat: implements the OAuth token exchange spec based on rfc8693

bojeil-google opened the pull request:
Implements an internal utility for exchanging OAuth tokens using the rfc/8693 spec.

bojeil-google committed:
feat: implements the OAuth token exchange spec based on rfc8693

JustinBeckwith commented on src/auth/stscredentials.ts:
```diff
@@ -0,0 +1,179 @@
+ * https://tools.ietf.org/html/rfc8693#section-2.1
+ */
+export interface StsCredentialsOptions {
+  grantType: string;
```
I don't know if the spec lays these out clearly, but I like to pull whatever text they have (if of appropriate value and length) to include in comments above the interface properties.  Like:

```js
/**
 * REQUIRED.  The value "urn:ietf:params:oauth:grant-type:token-exchange" indicates that a token exchange is being performed.
*/
grantType: string;

bojeil-google replied:
Done

JustinBeckwith commented on src/auth/stscredentials.ts:
```diff
@@ -0,0 +1,179 @@
+  actor_token_type?: string;
+  client_id?: string;
+  client_secret?: string;
+  [key: string]: string | undefined;
```
That is less than ideal :/  It effectively means any string key is value here with a string value.  Is this really an open property bag?

bojeil-google replied:
Ok. Currently there is only one non-standard field here. I manually added it.

JustinBeckwith commented on src/auth/stscredentials.ts:
```diff
@@ -0,0 +1,179 @@
+        opts
+      );
+      // Successful response.
+      const stsSuccessfulResponse = response.data;
```
I'm curious - why return the `response` object along with the data?

bojeil-google replied:
Mostly following the existing pattern established [here](https://github.com/bojeil-google/google-auth-library-nodejs/blob/a7e5701a8394d79fe93d28794467747a23cf9ff4/src/auth/oauth2client.ts#L309) and elsewhere in that file.

JustinBeckwith commented on src/auth/stscredentials.ts:
```diff
@@ -0,0 +1,179 @@
+    } catch (error) {
+      // Translate error to OAuthError.
+      if (error.response) {
+        throw getErrorFromOAuthErrorResponse(
```
Just to double check - does this method create a `new Error`, or modify the message on the existing?  I want to be about not creating a new error, because we want to preserve the original stack trace.

bojeil-google replied:
It was creating a new error and not preserving the original error data. I have extended this to preserve the original error data (including stack) but to modify the error message.

JustinBeckwith reviewed (approved):
LGTM with a few nits

grayside committed:
refactor(samples): idtoken-cloudrun => idtoken-serverless (#1025)

bojeil-google committed:
Addresses review comments.

Extends `getErrorFromOAuthErrorResponse` to preserve the original error properties.

yoshi-automation committed:
docs: README updated with new sample paths

bcoe commented:
@bojeil-google if you rebase with `master`  I believe it will address the docs failures you're seeing.

bojeil-google committed:
Merge branch 'master' of https://github.com/googleapis/google-auth-library-nodejs into sts-spec
